src/events/zookeeper.py

```diff
@@ -87,6 +87,17 @@ def _on_zookeeper_changed(self, event: RelationChangedEvent) -> None:
             for username, password in internal_user_credentials:
                 self.charm.state.cluster.update({f"{username}-password": password})
 
+        # Kafka keeps a meta.properties in every log.dir with a unique ClusterID
+        # this ID is provided by ZK, and removing it on relation-changed allows
+        # re-joining a ZK cluster undergoing backup restoration.
+        self.charm.workload.exec(
+            [
+                "bash",
+                "-c",
+                f"""find {self.charm.workload.paths.data_path} -type f -name meta.properties -delete || true""",
+            ]
+        )
+
```

Comment:
**todo:** AFAIK, this will run EVERY time ZooKeeper scales, is that correct? Do we want this behavior?

Reply:
This is needed for the "restore-while-related" feature because the ZK cluster ID changes when we restore. We can be a little smarter by checking the event diff (an ongoing restore removes only the "endpoints" field from the databag, so we could add a condition on that). I'll have a look

Reply:
I added a simple condition in 90c2f40c7dc3448ea6e50fd5a71ded438ee8d1da, let me know what you think

Reply:
You're going to hate me for suggesting this, but I actually think this needs to be part of the `zookeeper_client` database-blahblah lib with a custom event.

We should have some sort of flag that ZooKeeper writes to the relation, let's say `restoring=true`. Then, the lib will 'catch' the setting of `restoring=true` and emit on Kafka a `do_stuff_on_restore` event, that we can handle in `events/zookeeper.py` and do this behavior.

Reply:
I think any other behavior is likely to cause issues, as it's not just Kafka that will require ZooKeeper down the line.

Reply:
> We should have some sort of flag that ZooKeeper writes to the relation, let's say restoring=true. 

If you are ok with this, I will close this thread in a new issue because it entails several out-of-scope things:
- Using the database lib more → absolutely. We also have a comment saying that in the relation-changed handler
- Adding `restoring=True` → the charm-relation-interface PR is still open, so we could just do that. However, the databag content will then be different from the rest of our products. I don't know if it's an issue, but the discussion itself should not be taken lightly in a PR thread

Reply:
Agreed on a new issue being the right call.
On the `restoring=True` thing, FWIW I don't think we need to include EVERY flag in the relation-interface. I might be wrong on that, but I think it's important to know the 'valuable' ones, not ones that are just events. Maybe, not sure who the right person is to check with that.

Reply:
Actually, if we restrict the removal, we need to move this code above the early returns, otherwise it's never executed

Reply:
Changed in 502e313